istioctl/pkg/writer/compare/sds/util.go

```diff
@@ -0,0 +1,226 @@
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"time"
+
+	"istio.io/istio/istioctl/pkg/util/configdump"
+	"istio.io/istio/security/pkg/nodeagent/sds"
+	"istio.io/pkg/log"
+)
+
+// SecretItemDiff represents a secret that has been diffed between nodeagent and proxy
+type SecretItemDiff struct {
+	Agent string `json:"agent"`
+	Proxy string `json:"proxy"`
+	SecretItem
+}
+
+// SecretItem is an intermediate representation of secrets, used to provide a common
+// format between the envoy proxy secrets and node agent output which can be diffed
+type SecretItem struct {
+	Name        string `json:"resource_name"`
+	Data        string `json:"cert"`
+	Source      string `json:"source"`
+	Destination string `json:"destination"`
+	SecretMeta
+}
+
+// SecretMeta holds selected fields which can be extracted from parsed x509 cert
+type SecretMeta struct {
+	Valid        bool   `json:"cert_valid"`
+	SerialNumber string `json:"serial_number"`
+	Expires      string `json:"expires"`
+	ValidFrom    string `json:"valid_from"`
+	Type         string `json:"type"`
+}
+
+// NewSecretItemBuilder returns a new builder to create a secret item
+func NewSecretItemBuilder() SecretItemBuilder {
+	return &secretItemBuilder{}
+}
+
+// SecretItemBuilder wraps the process of setting fields for the SecretItem
+// and builds the Metadata fields from the cert contents behind the scenes
+type SecretItemBuilder interface {
+	Name(string) SecretItemBuilder
+	Data(string) SecretItemBuilder
+	Source(string) SecretItemBuilder
+	Destination(string) SecretItemBuilder
+	Build() (SecretItem, error)
+}
+
+// secretItemBuilder implements SecretItemBuilder, and acts as an intermediate before SecretItem generation
+type secretItemBuilder struct {
+	name   string
+	data   string
+	source string
+	dest   string
+	SecretMeta
+}
+
+// Name sets the name field on a secretItemBuilder
+func (s *secretItemBuilder) Name(name string) SecretItemBuilder {
+	s.name = name
+	return s
+}
+
+// Data sets the data field on a secretItemBuilder
+func (s *secretItemBuilder) Data(data string) SecretItemBuilder {
+	s.data = data
+	return s
+}
+
+// Source sets the source field on a secretItemBuilder
+func (s *secretItemBuilder) Source(source string) SecretItemBuilder {
+	s.source = source
+	return s
+}
+
+// Destination sets the destination field on a secretItemBuilder
+func (s *secretItemBuilder) Destination(dest string) SecretItemBuilder {
+	s.dest = dest
+	return s
+}
+
+// Build takes the set fields from the builder and constructs the actual SecretItem
+// including generating the SecretMeta from the supplied cert data, if present
+func (s *secretItemBuilder) Build() (SecretItem, error) {
+	result := SecretItem{
+		Name:        s.name,
+		Data:        s.data,
+		Source:      s.source,
+		Destination: s.dest,
+	}
+
+	var meta SecretMeta
+	var err error
+	if s.data != "" {
+
+		meta, err = secretMetaFromCert([]byte(s.data))
+		if err != nil {
+			log.Debugf("failed to parse secret resource %s from source %s: %v",
```

Comment:
why not return error here?

Reply:
Since we have the `Valid` field in the `SecretItem`, its actually acceptable for the secret not to parse here: we just mark `Valid` as false (since we need to display even secrets w/ invalid contents).